interp/src/interpreter.rs

```diff
@@ -0,0 +1,619 @@
+    /// Adds an update to the update queue; TODO; ok to drop prev and next?
+    pub fn add_update(
+        &mut self,
+        ucell: ir::Id,
+        uinput: Vec<ir::Id>,
+        uoutput: Vec<ir::Id>,
+        uvars: HashMap<String, u64>,
+    ) {
+        //println!("add update!");
+        let update = Update {
+            cell: ucell,
+            inputs: uinput,
+            outputs: uoutput,
+            vars: uvars,
+        };
+        self.update_queue.push(update);
+    }
+
+    /// Convenience function to remove a particular cell's update from the update queue
+    pub fn remove_update(&mut self, ucell: &ir::Id) {
+        self.update_queue.retain(|u| u.cell != ucell);
+    }
+
+    // TODO: should the return type be FuTIlResult<Environment>?
+    /// Simulates a clock cycle by executing the stored updates.
+    pub fn do_tick(mut self) -> Self {
+        let uq = self.update_queue.clone();
+        for update in uq {
+            let updated = update_cell_state(
+                &update.cell,
+                &update.inputs,
+                &update.outputs,
+                &self,
+            );
+            match updated {
+                Ok(updated_env) => {
+                    let updated_cell = updated_env
+                        .map
+                        .get(&update.cell)
+                        .unwrap_or_else(|| panic!("Can't get map"))
+                        .clone();
+                    self.map.insert(update.cell.clone(), updated_cell);
+                }
+                _ => panic!("Could not apply update. "),
+            }
+        }
+        self
+    }
+
+    /// Gets the cell based on the name; similar to find_cell in component.rs
+    fn get_cell(&self, cell: &ir::Id) -> Option<ir::RRC<ir::Cell>> {
+        self.cells
+            .values()
+            .find(|&g| g.borrow().name == *cell)
+            .map(|r| Rc::clone(r))
+    }
+
+    /// Outputs the cell state; TODO (write to a specified output in the future)
+    pub fn cell_state(&self) {
+        let state_str = self
+            .map
+            .iter()
+            .map(|(cell, ports)| {
+                format!(
+                    "{}\n{}",
+                    cell,
+                    ports
+                        .iter()
+                        .map(|(p, v)| format!("\t{}: {}", p, v))
+                        .collect::<Vec<_>>()
+                        .join("\n")
+                )
+            })
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        println!("{}\n{}\n{}", "=".repeat(30), state_str, "=".repeat(30))
+    }
+}
+
+/// Evaluates a group, given an environment.
+pub fn eval_group(
+    group: ir::RRC<ir::Group>,
+    env: Environment,
+) -> FutilResult<Environment> {
+    eval_assigns(&(group.borrow()).assignments, &env)
+}
+
+// XXX(karen): I think it will need another copy of environment for each iteration of assignment statements
+/// Evaluates a group's assignment statements in an environment.
```

Comment:
can you write a description of broad algorithm used here? this would make the code easier to follow

Reply:
These were originally added in as comments in the body of the function, but I've added a description there now.